cli: add CLI flag to enable DRPC for init and node commands

cthumuluru-crdb opened the pull request:
DRPC is controlled through a cluster setting and is disabled by default. However, CLI commands such as init and debug run without access to cluster settings. To allow these commands to use DRPC, this change introduces a new `--use-new-rpc` flag. This change adds the flag to the init and node commands.

Fixes: #155593, #155596
Epic: CRDB-55200
Release note: None

cthumuluru-crdb committed:
*: set drpc cluster setting based on CLI flag

Set `rpc.experimental_drpc.enabled=true` when the
`--use-new-rpc` CLI flag is provided instead of maintaining it
separately in server config. Since majority of the code paths use
cluster settings to determine whether drpc is enabled, this simplifies
the code.

Fixes: #155593
Epic: CRDB-55200
Relese note: None

cthumuluru-crdb committed:
cli,rpc: add drpc support for init command

Init command can now use drpc for inter-node communication by
providing `--use-new-rpc` flag.

Fixes: #155593
Epic: CRDB-55200
Release note: None

cthumuluru-crdb committed:
cli: add drpc support for node commands

Decommission, Recommission and Drain commands can now use drpc for
inter-node communication by providing `--use-new-rpc` flag.

Fixes: #155596
Epic: CRDB-55200
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/159246)
<!-- Reviewable:end -->

cthumuluru-crdb commented on pkg/cli/start.go:
Is your question related to adding a CLI flag or updating the default cluster settings using the flag value?

shubhamdhama replied:
The latter.

cthumuluru-crdb replied:
Alternatives considered is to use a config parameter. Since all the code paths including tests use cluster settings to check if DRPC is enabled or not, it looked natural to override the setting with flag value. Do you see any challenges with it?

cthumuluru-crdb replied:
@stevendanna, do you see any issues with overriding the default value based on the cli flag?

stevendanna replied:
Discussion notes:

If we want to get rid of this setting, it might be better to do the work to thread the setting through to some configuration item and remove the setting.

If we don't want to remove the setting, I'd probably still thread through the configuration and then implement the priority in a simple function.

cthumuluru-crdb committed:
cli: add drpc support for debug info commands

Debug commands that retrieve cluster information: debug gossip-values, debug
tsdump, debug zip, and debug list-files can now use drpc for inter-node
communication by providing `--use-new-rpc` flag.

Informs: #155598
Epic: CRDB-55200
Release note: None

cthumuluru-crdb committed:
cli: add drpc support for debug operation commands

Debug commands that perform cluster operations: debug send-kv-batch and debug
reset-quorum can now use drpc for inter-node communication by providing
`--use-new-rpc` flag.

Informs: #155598
Epic: CRDB-55200
Release note: None

cthumuluru-crdb committed:
cli: add drpc support for debug recover commands

Debug recover commands: debug recover collect-info, debug recover plan, debug
recover execute, and debug recover verify can now use drpc for inter-node
communication by providing `--use-new-rpc` flag.

Fixes: #155598
Epic: CRDB-55200
Release note: None

cthumuluru-crdb committed:
cli: add drpc support for gen haproxy command

This command uses RPC connections to retrieve cluster node information for
generating HAProxy configuration. It connects to the cluster, creates a
`StatusClient`, and makes RPC calls to get node status and liveness
information. `gen haproxy command` can now use drpc for inter-node
communication by providing `--use-new-rpc` flag.

Informs: #155599
Epic: CRDB-55200
Release note: None
